submariner/globalnet-enhancement2-0.md

```diff
@@ -0,0 +1,197 @@
+   annotated with a globalIp.
+5. For Headless Services, necessary ingress rules will be programmed to support connectivity to
+   the backend Pods. In a Headless Service use-case, every backend Pod will be annotated with
+   a unique globalIP and the same globalIP will be used for both Ingress as-well-as Egress.
+   globalIPs allocated for Pods backed by Headless Services will have highest priority over other
+   egress policies.
+
+![GlobalnetTopology](./images/globalnet-topology.png)
+
+If the external application in a remote Cluster is interested in consuming the exported
+Service, it will have to reach the globalIP assigned to the exportedService and the response
+from the backend Pods will have the source as the globalIP of the exportedService.
+
+On the other hand, when the backend Pod tries to talk to the external application in remote
+Cluster, the external application will see that traffic is coming with a globalIP that is
+assigned either at the Cluster level, or namespace level, or the one requested by the user
+at Pod level. This asynchronous behavior will be analogous to the behavior seen with standard
+K8s applications when they talk to any Services. Applications desiring to have the same IPaddress
+for both ingress and egress traffic can create Headless Services in front of them.
+
+The proposal has its own set of Pros and Cons.
+
+## Pros
+
+Since we will be annotating only select K8s objects with globalIp, it will help in improved scalability.
+Also, this translates to fewer iptable rules on the active Gateway node of the cluster which will
+improve the overall performance of the solution.
+
+In the current implementation, when a client Pod is scheduled, until it's annotated with a globalIp,
+and the corresponding egress rules are programmed on the node, it was not able to communicate with
+remote Services. This was creating some issues in-terms of user-experience. In this proposal, as we
+have a globalIp at the cluster level, a Pod can communicate with remote Services right away which
+will be closer to the Vanilla Submariner experience.
+
+## Cons
+
+A limitation with the proposed approach is that we will lose the unique identity for **every**
+Pod which is supported today. For users that intend to use Network Policies across Clusters or
+for debugging purposes, this can be mitigated by creating the necessary CRDs which create unique
+globalIps at the Namespace/Pod level.
+
+The following CRD is proposed to support the new design.
+
+```Go
+type GlobalnetEgressIP struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec               EgressIPSpec `json:"spec"`
+}
+
+type EgressIPSpec struct {
+    // Selects the pods to which the GlobalnetEgressIP will be applied.
+    // If multiple pods match the selector the same egressIP will be used for all those pods.
+    // +optional
+    PodSelector PodSelectorSpec `json:"podSelector,omitempty"`
```

Comment:
Let's use the standard kubernetes selectors instead:
```
k8s.io/apimachinery/pkg/apis/meta/v1.LabelSelector
```
https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/types.go#L1081

This will allow use to use MatchExpressions for free... the implementation of the matching is given to us.

Example of matching:
https://github.com/submariner-io/coastguard/blob/devel/pkg/networkpolicy/networkpolicy.go#L147

Reply:
Yes, I saw this but I intentionally considered the PodSelector. 

IMHO, I think we don't really need the complex LabelSelector for our use-case. Network Policies might have their reasons, but GlobalnetEgressIP CRD is more like a Service that is backed by Pods. Service CRD also uses PodSelector. What do you say?

Reply:
[This discussion](https://github.com/kubernetes/kubernetes/issues/48528) is related? I think that most of the use cases would be covered by the same interface to Service, but some might find useful to the "complex" selector, as there seems such a use case even with Service.

Reply:
That the LabelSelector brings greater flexibility to the user, without (IMO at least) greater cost of development, or worsened UI.

The fact that ServiceSelector uses just a map lot labels I believe it has to do with it being much older (and more core) than network policies, which makes it harder to upgrade. In this case it's a fresh start.

Reply:
Done